code/controllers/subsystem/input.dm

```diff
@@ -29,12 +42,53 @@ SUBSYSTEM_DEF(input)
 	)
 
 // Badmins just wanna have fun ♪
-/datum/controller/subsystem/input/proc/refresh_client_macro_sets()
+/datum/controller/subsystem/verb_manager/input/proc/refresh_client_macro_sets()
 	var/list/clients = GLOB.clients
 	for(var/i in 1 to clients.len)
 		var/client/user = clients[i]
 		user.set_macros()
 
-/datum/controller/subsystem/input/fire()
-	for(var/mob/user as anything in GLOB.keyloop_list)
-		user.focus?.keyLoop(user.client)
+/datum/controller/subsystem/verb_manager/input/can_queue_verb(datum/callback/verb_callback/incoming_callback, control)
+	//make sure the incoming verb is actually something we specifically want to handle
+	if(control != "mapwindow.map")
+		return FALSE
+
+	if(average_click_delay >= MAXIMUM_CLICK_LATENCY || !..())
+		current_clicks++
+		average_click_delay = MC_AVG_FAST_UP_SLOW_DOWN(average_click_delay, 0)
+		return FALSE
+
+	return TRUE
+
+///stupid workaround for byond not recognizing the /atom/Click typepath for the queued click callbacks
+/atom/proc/wrap_Click(location, control, params)
+	if(usr)
+		Click(location, control, params)
+
+
+/datum/controller/subsystem/verb_manager/input/fire()
+	var/moves_this_run = 0
+	var/deferred_clicks_this_run = 0 //acts like current_clicks but doesnt count clicks that dont get processed by SSinput
+
+	for(var/datum/callback/verb_callback/queued_click in verb_queue)
+		average_click_delay = MC_AVG_FAST_UP_SLOW_DOWN(average_click_delay, (REALTIMEOFDAY - queued_click.creation_time) SECONDS)
+		queued_click.InvokeAsync()
+
+		current_clicks++
+		deferred_clicks_this_run++
+
```

Comment:
```suggestion
	for(var/datum/callback/verb_callback/queued_click in verb_queue)
		average_click_delay = MC_AVG_FAST_UP_SLOW_DOWN(average_click_delay, (REALTIMEOFDAY - queued_click.creation_time) SECONDS)
		queued_click.InvokeAsync()

	current_clicks += length(verb_queue)
	deferred_clicks_this_run += length(verb_queue)

```

Reply:
unless this is to filter nulls in which case uh, how

Reply:
it is meant to filter elements of verb_queue that arent iterated over. i changed the for loop to stack_trace() and continue if it goes over something that shouldnt be there.